tilavarauspalvelu/models/user/actions.py

```diff
@@ -187,3 +194,105 @@ def can_anonymize(self) -> UserAnonymizationInfo:
             has_open_applications=has_open_applications,
             has_open_payments=has_open_payments,
         )
+
+    def get_ad_group_roles(self) -> dict[UserRoleChoice, set[int]]:
+        """
+        Parse information from user's AD groups for creating unit roles.
+
+        AD groups that give role should be of format `<prefix>__varaamo__<roles>__<tprek_id>` where:
+         - '<prefix>': can be anything (usually identifies the unit)
+         - '<roles>': roles matching `UserRoleChoice` the user should have (separated by "__")
+         - '<tprek_id>': `Unit.tprek_id` for the unit the role is for.
+        """
+        identifier = "__VARAAMO__"
+        ad_group_names: set[str] = set(
+            self.user.ad_groups.filter(name__icontains=identifier)
+            .annotate(upper_name=Upper("name"))
+            .values_list("upper_name", flat=True)
+        )
+
+        if self.user.is_superuser and settings.FAKE_SUPERUSER_AD_GROUPS:
+            ad_group_names.update(name.upper() for name in settings.FAKE_SUPERUSER_AD_GROUPS)
+
+        units_by_role: dict[UserRoleChoice, set[int]] = {}
+
+        for ad_group_name in ad_group_names:
+            parts = ad_group_name.split(identifier, maxsplit=1)
+            if len(parts) != 2:  # noqa: PLR2004
+                continue
+
+            info_parts = parts[1].split("__")
+            if len(info_parts) < 2:  # noqa: PLR2004
+                continue
+
+            tprek_id = info_parts[-1]
+            unit = Unit.objects.filter(tprek_id=tprek_id).first()
+            if unit is None:
+                continue
+
+            if not unit.allow_permissions_from_ad_groups:
+                continue
+
+            for role in info_parts[:-1]:
+                if role not in UserRoleChoice.allowed_roles_for_ad_permissions():
+                    continue
+
+                user_role = UserRoleChoice(role)
+                units_by_role.setdefault(user_role, set()).add(unit.id)
+
+        return units_by_role
+
+    def update_unit_roles_from_ad_groups(self) -> None:
+        current_ad_roles = self.get_ad_group_roles()
+        existing_ad_roles = {
+            UserRoleChoice(role.role): role  #
```

Comment:
The `.filter(from_ad_group=True)` in the queryset only selects UnitRoles from AD groups and since this is the only place those should be created from, and here we make sure that there is only one UnitRole for each role, this shouldn't remove anything. I should probably add a constraint to the model just in case.

Reply:
Added a constraint an a comment explaining this.